js/woo.js

```diff
@@ -0,0 +1,1532 @@
+        }
+        return this.safeTrade ({
+            'id': id,
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'symbol': symbol,
+            'side': side,
+            'price': price,
+            'amount': amount,
+            'cost': cost,
+            'order': order_id,
+            'takerOrMaker': takerOrMaker,
+            'type': undefined,
+            'fee': {
+                'cost': feeValue,
+                'currency': feeAsset,
+            },
+            'info': trade,
+        }, market);
+    }
+
+    async fetchCurrencies (params = {}) {
+        // TODO: we need to write them to merge 'token' and 'token_network' objects from API, as it's horrificly bad atm.
+        let method = undefined;
+        const [ marketType, query ] = this.handleMarketTypeAndParams ('fetchCurrencies', undefined, params);
+        method = this.getSupportedMapping (marketType, {
+            'spot': 'v1PublicGetToken',
+        });
+        const response1 = await this[method] (query);
+        //
+        // {
+        //     rows: [
+        //         {
+        //             token: "ETH_USDT",
+        //             fullname: "Tether",
+        //             decimals: 6,
+        //             balance_token: "USDT",
+        //             created_time: "0",
+        //             updated_time: "0"
+        //         },
+        //         {
+        //             token: "BSC_USDT",
+        //             fullname: "Tether",
+        //             decimals: 18,
+        //             balance_token: "USDT",
+        //             created_time: "0",
+        //             updated_time: "0"
+        //         },
+        //         {
+        //             token: "ZEC",
+        //             fullname: "ZCash",
+        //             decimals: 8,
+        //             balance_token: "ZEC",
+        //             created_time: "0",
+        //             updated_time: "0"
+        //         },
+        //         ...
+        //     ],
+        //     success: true
+        // }
+        //
+        const result = {};
+        const data1 = this.safeValue (response1, 'rows', []);
+        const derivedCurrenciesData = {};
+        for (let i = 0; i < data1.length; i++) {
+            const item = data1[i];
+            const tokenCode = this.safeString (item, 'balance_token');
+            const code = this.safeCurrencyCode (tokenCode);
+            if (!(code in derivedCurrenciesData)) {
+                derivedCurrenciesData[code] = {
+                    'fullname': this.safeString (item, 'fullname'),
+                    'networks': {},
+                };
+            }
+            const chainedTokenCode = this.safeString (item, 'token');
+            const parts = chainedTokenCode.split ('_');
+            const networkId = parts.length === 2 ? this.safeString (parts, 0) : chainedTokenCode;
```

Comment:
This is a violation of two CONTRIBUTING guidelines:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#working-with-array-lengths
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#using-ternary-conditionals